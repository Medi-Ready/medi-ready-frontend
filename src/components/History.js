import styled from "styled-components";
import { useQuery } from "react-query";
import React, { useContext, useEffect, useState } from "react";

import { getPrescriptions } from "../api";
import { ModalContext } from "../contexts/ModalContext";

import Detail from "./Detail";
import Pagination from "./Pagination";
import Loading from "./Shared/Loading";
import HistoryListItem from "./Shared/HistoryListItem";

const History = ({ queryClient }) => {
  const { handleModal } = useContext(ModalContext);
  const [page, setPage] = useState(0);

  const { data, isPreviousData, isFetching, isStale } = useQuery(["prescriptions", page],
    () => getPrescriptions(page), { keepPreviousData: true, staleTime: 5 * 1000 },
  );

  useEffect(() => {
    if (data?.hasMoreData && isStale) {
      queryClient.prefetchQuery(["prescriptions", page + 1], () => getPrescriptions(page + 1));
    }
  }, [data, page, queryClient]);

  const ViewDetail = (prescription) => {
    handleModal(<Detail prescription={prescription} />);
  };

  return (
    <>
      <h2>Prescription History</h2>

      <Wrapper>
        <ul>
          <li className="name-card">
            <b>Name</b>
            <span>Status</span>
            <span>Date</span>
          </li>
        </ul>
        <ul className="list-items">
          {isFetching ? (
            <Loading />
          ) : (
            data.prescriptions.map((prescription) => {
              return (
                <HistoryListItem
                  key={prescription.prescription_id}
                  prescription={prescription}
                  openModal={() => ViewDetail(prescription)}
                />
              );
            })
          )}
        </ul>
      </Wrapper>
      <Pagination
        page={page}
        setPage={setPage}
        hasMoreData={data?.hasMoreData}
        isPreviousData={isPreviousData}
      />
    </>
  );
};

const Wrapper = styled.div`
  margin-top: 20px;
  min-height: 60vh;
  padding: ${({ theme }) => theme.padding.default};
  border-radius: 10px;
  background-color: ${({ theme }) => theme.color.white};
  box-shadow: rgb(149 157 165 / 20%) 0px 8px 24px;

  ul {
    padding-top: 5px;
  }

  ul:first-child {
    border-bottom: 1px solid #e9e9e9;
  }

  li {
    display: flex;
    justify-content: space-between;
    align-items: center;
    border-bottom: 1px solid #eee;
    padding: 30px 15px;
    height: 50px;
  }

  li.name-card b {
    width: 20%;
    font-size: 16px;
    color: #767676;
  }

  li.name-card span {
    color: #767676;
    font-size: 16px;
  }

  li span:nth-child(2) {
    width: 50%;
  }

  li span:nth-child(3) {
    width: 15%;
  }

  li:not(.name-card):hover {
    cursor: pointer;
  }
`;

export default History;
