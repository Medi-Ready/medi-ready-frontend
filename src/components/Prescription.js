import React from "react";

const Prescription = () => {
  return (
    <>
      <h2>Prescription</h2>
    </>
  );
};

export default Prescription;
