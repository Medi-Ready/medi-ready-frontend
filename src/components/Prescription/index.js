import { useQuery, useMutation } from "react-query";
import React, { useState, useEffect, useContext, useCallback } from "react";

import { debounce } from "lodash";
import styled from "styled-components";
import { getMedicineNames, getQueue, postMedicine, postPrescription } from "../../api";

import { ModalContext } from "../../contexts/ModalContext";

import Queue from "./Queue";
import Badge from "../Shared/Badge";
import SearchBar from "../SearchBar";
import Button from "../Shared/Button";
import FlexBox from "../Shared/FlexBox";
import Checkbox from "../Shared/Checkbox";
import TextArea from "../Shared/TextArea";
import TextInput from "../Shared/TextInput";
import ConfirmModal from "../Shared/Modal/ConfirmModal";

const Prescription = () => {
  const [message, setMessage] = useState("");
  const [keyword, setKeyword] = useState("");
  const [medicine, setMedicine] = useState("");
  const [duration, setDuration] = useState("");
  const [description, setDescription] = useState("");
  const [errorMessage, setErrorMessage] = useState("");
  const [medicineList, setMedicineList] = useState([]);
  const [doseTimeList, setDoseTimeList] = useState([]);
  const [searchResult, setSearchResult] = useState([]);
  const [targetUserInfo, setTargetUserInfo] = useState({});
  const [isPrescriptionSubmit, setIsPrescriptionSubmit] = useState(false);

  const { name, picture } = targetUserInfo;
  const { handleModal } = useContext(ModalContext);

  useEffect(() => {
    setDoseTimeList(doseTimeDataList);
  }, []);

  useEffect(() => {
    setErrorMessage("");
  }, [duration, description, targetUserInfo, medicineList, doseTimeList]);

  const onChangeHanlder = (event, setState) => {
    setState(event.target.value);
  };

  const { data, isLoading } = useQuery("queue", getQueue, {
    refetchInterval: 3000,
  });

  const medicineNamesMutation = useMutation(getMedicineNames, {
    onSuccess: (result) => {
      let { data } = result;

      if (data) {
        data = data.slice(0, 5);
        setSearchResult(data);
      }
    },
  });

  const handleKeywordChange = (newKeyword) => {
    setKeyword(newKeyword);

    if (newKeyword) {
      searchMedicines(newKeyword);
    }
  };

  const searchMedicines = useCallback(
    debounce((newValue) => {
      medicineNamesMutation.mutate({ keyword: newValue });
    }, 300),
    [],
  );

  const medicineMutation = useMutation(postMedicine, {
    onSuccess: (result) => {
      const { data } = result;

      if (data) {
        setMedicine(data);
        setKeyword("");
        setSearchResult([]);
        setMedicineList(prevItems => [...prevItems, {
          id: data.medicine_id,
          name: data.name,
        }]);
      } else {
        setErrorMessage("해당 약은 존재하지 않습니다.");
      }
    },
  });

  const prescriptionMutation = useMutation(postPrescription, {
    onSuccess: ({ result }) => {
      if (result === "success") {
        setMessage("처방전 전송이 완료되었습니다.");
      }
    },
  });

  const handleSearch = (event) => {
    event.preventDefault();

    const { search } = event.target;

    medicineMutation.mutate({ name: event.target.search.value });

    event.target.search.value = "";
  };

  const handleChange = (event) => {
    const { id, checked } = event.target;

    if (id === "all") {
      let allCheckedList = doseTimeList.map((time) => {
        return { ...time, isChecked: checked };
      });

      setDoseTimeList(allCheckedList);
    } else {
      let checkedList = doseTimeList.map((time) =>
        time.id === id ? { ...time, isChecked: checked } : time,
      );

      setDoseTimeList(checkedList);
    }
  };

  const handleSubmit = (event) => {
    event.preventDefault();

    const doseTimes = [];

    for (const doseTime of doseTimeList) {
      if (doseTime.isChecked) {
        doseTimes.push(doseTime.id);
      }
    }

    const medicineIdList = medicineList.map(({ id }) => id);

    handleModal(<ConfirmModal setIsPrescriptionSubmit={setIsPrescriptionSubmit} />);

    const prescriptionForm = {
      doseTimes,
      duration,
      description,
      medicines: medicineIdList,
      date: new Date().toISOString(),
      patient_id: targetUserInfo.patient_id,
    };

    if (isPrescriptionSubmit) {
      prescriptionMutation.mutate(prescriptionForm);
      setDuration("");
      setDescription("");
      setErrorMessage("");
      setMedicineList([]);
      setTargetUserInfo({});
      setIsPrescriptionSubmit(false);
      setDoseTimeList(doseTimeDataList);
    }
  };

  const doseTimeDataList = [
    { id: "morning", label: "아침", name: "doseTime" },
    { id: "lunch", label: "점심", name: "doseTime" },
    { id: "dinner", label: "저녁", name: "doseTime" },
    { id: "beforeBed", label: "취침전", name: "doseTime" },
  ];

  return (
    <>
      <FlexBox>
        <h2>Prescription</h2>
      </FlexBox>

      <Wrapper>
        <FlexBox>
          <div>
            <UserInfo>
              {name && (
                <>
                  <span>
                    <img src={picture} alt={name} />
                  </span>
                  <b>{name}</b>
                  <Badge color="green">Treating</Badge>
                </>
              )}
            </UserInfo>

            <form onSubmit={handleSearch} autocomplete="off">
              <SearchBox>
                <SearchBar
                  type="text"
                  name="search"
                  label="search"
                  keyword={keyword}
                  results={searchResult}
                  setValue={handleKeywordChange}
                  placeholder="Enter Medicine Name"
                />
                <SearchBarButton type="submit" text="Search">추가</SearchBarButton>
                <div>
                  {medicineList.map((item, i) => (
                    <TextInput
                      key={item.id, i}
                      label={`medicine${i}`}
                      value={item.name.slice(0, 35) + "…"}
                      name="medicine"
                      disabled
                    />
                  ))}
                </div>
              </SearchBox>
            </form>
          </div>

          <div>
            <CheckboxList>
              <Checkbox
                id="all"
                name="all"
                label="전체"
                checked={doseTimeList.filter(time => time?.isChecked !== true).length < 1}
                onChange={handleChange}
              />
              {doseTimeList.map((time) => {
                const { id, name, label } = time;

                return (
                  <Checkbox
                    checked={time?.isChecked || false}
                    id={id}
                    name={name}
                    label={label}
                    onChange={handleChange}
                  />
                );
              })}
            </CheckboxList>
          </div>
        </FlexBox>

        <form onSubmit={handleSubmit}>
          <FlexBox>
            <div>
              <TextArea name="description" onChange={(event) => onChangeHanlder(event, setDescription)}></TextArea>
            </div>
            <InputButtonBox>
              <TextInput width="60px" label="duration" name="duration" onChange={(event) => onChangeHanlder(event, setDuration)} />
              <span>일치</span>
              <Button type="submit">처방</Button>
            </InputButtonBox>
          </FlexBox>
        </form>
        {message && <p>처방전 전송이 완료되었습니다.</p>}
        {errorMessage && <ErrorMessage>{errorMessage}</ErrorMessage>}
      </Wrapper>

      {!isLoading && (
        <Wrapper>
          <Queue
            queue={data}
            targetUser={targetUserInfo}
            setTargetUserInfo={setTargetUserInfo} />
        </Wrapper>
      )}
    </>
  );
};

const Wrapper = styled.div`
  margin-top: 20px;
  padding: 20px 20px 35px;
  border-radius: 10px;
  background-color: ${({ theme }) => theme.color.white};
  box-shadow: rgb(149 157 165 / 20%) 0px 8px 24px;
`;

const CheckboxList = styled.ul`
  display: inline-flex;
  flex-wrap: wrap;
  justify-content: space-between;
  max-width: 450px;
  padding-left: 15px;
  vertical-align: top;
`;

const SearchBarButton = styled(Button)`
  position: absolute;
  top: 0;
  right: 0;
`;

const UserInfo = styled.div`
  position: relative;
  padding-bottom: 20px;

  b {
    display: inline-block;
    margin-left: 12px;
    font-size: 18px;
    font-weight: 600;
  }

  span {
    display: inline-block;
    font-size: 12px;

    img {
        display: inline-block;
        width: 40px;
        object-fit: cover;
        border-radius: 50%;
        vertical-align: middle;
      }
  }

  em {
    position: absolute;
    top: 0;
    right: 0;
  }
`;

const SearchBox = styled.div`
  position: relative;

  button {
    position: absolute;
    right: 0;
  }

  div input {
    margin-top: 10px;
  }
`;

const InputButtonBox = styled.div`
  position: absolute;
  right: 0;
  bottom: 0;
  padding-left: 10px;

  > * {
    margin-left: 20px;
  }
`;

export default Prescription;
