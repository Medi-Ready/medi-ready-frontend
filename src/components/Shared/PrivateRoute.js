import React from "react";
import { Route, Redirect } from "react-router-dom";

const PrivateRoute = ({ component: Component, isAuthenticated, ...rest }) => {
  return (
    <Route
      {...rest}
      render={(props) =>
        isAuthenticated
          ? <Component {...props} {...rest} />
          : <Redirect to="/login" />
      }
    />
  );
};

export default PrivateRoute;
