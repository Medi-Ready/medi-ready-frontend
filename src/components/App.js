import React, { useState, useEffect } from "react";
import { Route, Switch, Redirect, useHistory } from "react-router-dom";
import { QueryClient, QueryClientProvider } from "react-query";
import styled, { ThemeProvider } from "styled-components";
import { ReactQueryDevtools } from "react-query/devtools";

import GlobalStyles from "../styles";
import theme from "../styles/theme";
import Header from "./Header";
import Footer from "./Footer";
import Navigation from "./Navigation";
import Login from "./Login";
import History from "./History";
import Settings from "./Settings";
import DashboardList from "./DashboardList";
import Prescription from "../components/Prescription";
import { ModalProvider } from "../contexts/ModalContext";
import { getAuthCheck } from "../api";

const queryClient = new QueryClient();

const App = () => {
  const [user, setUser] = useState("");
  const history = useHistory();

  useEffect(() => {
    const checkLoginStatus = async () => {
      const { result, data } = await getAuthCheck();

      if (result === "success") {
        setUser(data);
      } else {
        setUser("");
        history.push("/login");
      }
    };

    checkLoginStatus();
  }, []);

  return (
    <Wrapper>
      <ThemeProvider theme={theme}>
        <QueryClientProvider client={queryClient}>
          <GlobalStyles />
          <Navigation isLoggedIn={user} onLogout={setUser} />
          <Section>
            <Header />
            <Article>
              <ModalProvider>
                <Switch>
                  <Route path="/dashboard">
                    <DashboardList />
                  </Route>
                  <Route path="/prescription">
                    <Prescription />
                  </Route>
                  <Route path="/prescriptions">
                    <History />
                  </Route>
                  <Route path="/prescriptions/:id">

                  </Route>
                  <Route path="/settings">
                    <Settings />
                  </Route>
                  <Route path="/login">
                    <Login onSuccess={setUser} />
                  </Route>
                  <Route path="/" exact>
                    <Redirect to="/dashboard" />
                  </Route>
                </Switch>
              </ModalProvider>
            </Article>
            <Footer />
          </Section>
        </QueryClientProvider>
      </ThemeProvider>
    </Wrapper>
  );
};

const Wrapper = styled.div`
  display: flex;
`;

const Section = styled.section`
  width: calc(100% - 230px);
  min-height: 100vh;
  background-color: ${({ theme }) => theme.color.grey};
`;

const Article = styled.article`
  height: calc(100% - 115px);
  padding: ${({ theme }) => theme.padding.big};
`;

export default App;
