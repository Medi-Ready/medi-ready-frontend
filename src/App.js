import React, { useState, useEffect } from "react";
import { QueryClient, QueryClientProvider } from "react-query";
import { Route, Switch, Redirect, useHistory } from "react-router-dom";
import styled, { ThemeProvider } from "styled-components";

import { getAuthCheck } from "./api";
import { ModalProvider } from "./contexts/ModalContext";

import GlobalStyles from "./styles";
import theme from "./styles/theme";
import Header from "./components/Header";
import Footer from "./components/Footer";
import Navigation from "./components/Navigation";
import Login from "./pages/Login";
import Error from "./components/Shared/Error";
import History from "./pages/History";
import Settings from "./pages/Settings";
import Prescription from "./pages/Prescription";
import DashboardList from "./pages/Dashboard";
import PrivateRoute from "./components/Shared/PrivateRoute";

const queryClient = new QueryClient();

const App = () => {
  const history = useHistory();

  const [user, setUser] = useState("");
  const [isSignedIn, setIsSignedIn] = useState(false);

  useEffect(() => {
    const checkLoginStatus = async () => {
      try {
        const { data, result, pharmacyInformation } = await getAuthCheck();

        if (result === "success") {
          setUser({ ...data, ...pharmacyInformation });
          history.push("/");
        } else {
          setUser("");
        }

      } catch (error) {
        return <Error error={error} />;
      }
    };

    checkLoginStatus();
  }, [isSignedIn]);

  return (
    <Wrapper>
      <ThemeProvider theme={theme}>
        <QueryClientProvider client={queryClient}>
          <GlobalStyles />
          <Navigation isSignedIn={user} onLogout={setUser} />
          <Section>
            <Header userInfo={user} />
            <ModalProvider>
              <Switch>
                <PrivateRoute
                  path="/dashboard"
                  isAuthenticated={user}
                  component={DashboardList}
                  userInfo={user}
                />
                <PrivateRoute
                  path="/prescription"
                  isAuthenticated={user}
                  component={Prescription}
                  userInfo={user}
                />
                <PrivateRoute
                  path="/prescriptions"
                  isAuthenticated={user}
                  component={History}
                  queryClient={queryClient}
                />
                <PrivateRoute
                  path="/settings"
                  isAuthenticated={user}
                  component={Settings}
                  userInfo={user}
                />
                <Route path="/login">
                  <Login onSuccess={setUser} setIsSignedIn={setIsSignedIn} />
                </Route>
                <Route path="/error">
                  <Error />
                </Route>
                <Route path="/" exact>
                  <Redirect to="/dashboard" />
                </Route>
              </Switch>
            </ModalProvider>
            <Footer />
          </Section>
        </QueryClientProvider>
      </ThemeProvider>
    </Wrapper>
  );
};

const Wrapper = styled.div`
  display: flex;
`;

const Section = styled.section`
  width: calc(100% - 230px);
  min-height: 100vh;
  background-color: ${({ theme }) => theme.color.grey};
`;

export default App;
