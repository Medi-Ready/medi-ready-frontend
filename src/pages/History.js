import React, { useContext, useEffect, useState } from "react";
import { useQuery } from "react-query";
import styled from "styled-components";

import { getPrescriptions } from "../api";
import { ModalContext } from "../contexts/ModalContext";

import Error from "../components/Shared/Error";
import Detail from "../components/History/Detail";
import Loading from "../components/Shared/Loading";
import ListItem from "../components/History/ListItem";
import Pagination from "../components/Shared/Pagination";
import { PageTitle, PageContent } from "../components/Base";

const History = ({ queryClient }) => {
  const { handleModal } = useContext(ModalContext);
  const [page, setPage] = useState(0);

  const { data, error, isPreviousData, isFetching, isStale, isError } = useQuery(["prescriptions", page],
    () => getPrescriptions(page), { keepPreviousData: true, staleTime: 5 * 1000 },
  );

  useEffect(() => {
    if (data?.hasMoreData && isStale) {
      queryClient.prefetchQuery(["prescriptions", page + 1], () => getPrescriptions(page + 1));
    }
  }, [data, page, queryClient]);

  const ViewDetail = (prescription) => {
    handleModal(<Detail prescription={prescription} />);
  };

  if (isError) {
    return <Error error={error} />;
  }

  return (
    <PageContent>
      <PageTitle>Prescription History</PageTitle>

      <Wrapper>
        <ul>
          <li className="name-card">
            <b>이름</b>
            <span>복용지도</span>
            <span>처방전 발행날짜</span>
          </li>
        </ul>
        <ul className="list-items">
          {isFetching ? (
            <Loading />
          ) : (
            data.prescriptions?.map((prescription) => {
              return (
                <ListItem
                  key={prescription.prescription_id}
                  prescription={prescription}
                  openModal={() => ViewDetail(prescription)}
                />
              );
            })
          )}
        </ul>
      </Wrapper>
      <Pagination
        page={page}
        setPage={setPage}
        hasMoreData={data?.hasMoreData}
        isPreviousData={isPreviousData}
      />
    </PageContent>
  );
};

const Wrapper = styled.div`
  margin-top: 20px;
  min-height: 470px;
  padding: ${({ theme }) => theme.padding.default};
  border-radius: 10px;
  background-color: ${({ theme }) => theme.color.white};
  box-shadow: rgb(149 157 165 / 20%) 0px 8px 24px;

  ul {
    padding-top: 5px;
  }

  ul:first-child {
    border-bottom: 1px solid #e9e9e9;
  }

  li {
    display: flex;
    justify-content: space-between;
    align-items: center;
    border-bottom: 1px solid #eee;
    padding: 8px 10px;
  }

  li.name-card b {
    width: 20%;
    font-size: 14px;
    color: #767676;
  }

  li.name-card span {
    color: #767676;
    font-size: 14px;
  }

  li span:nth-child(2) {
    width: 50%;
  }

  li span:nth-child(3) {
    width: 15%;
  }

  li:not(.name-card):hover {
    cursor: pointer;
  }
`;

export default History;
