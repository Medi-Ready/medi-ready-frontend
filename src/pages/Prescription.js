import { useMutation } from "react-query";
import React, { useState, useEffect, useContext } from "react";
import styled from "styled-components";

import { postPrescription } from "../api";
import { ModalContext } from "../contexts/ModalContext";

import Queue from "../components/Prescription/Queue";

import Loading from "../components/Shared/Loading";
import FlexBox from "../components/Shared/FlexBox";
import Checkbox from "../components/Shared/Checkbox";
import UserInfo from "../components/Prescription/UserInfo";
import { PageTitle, PageContent } from "../components/Base";
import ErrorMessage from "../components/Shared/ErrorMessage";
import InputForm from "../components/Prescription/InputForm";
import SearchForm from "../components/Prescription/SearchForm";
import ConfirmMessage from "../components/Shared/Modal/ConfirmMessage";

const Prescription = () => {
  const { handleModal } = useContext(ModalContext);

  const [error, setError] = useState("");
  const [formData, setFormData] = useState({});
  const [isSubmit, setIsSubmit] = useState(false);
  const [medicineList, setMedicineList] = useState([]);
  const [selectedUser, setSelectedUser] = useState({});
  const [doseTimeList, setDoseTimeList] = useState([]);

  useEffect(() => {
    setDoseTimeList(doseTimeDataList);
  }, []);

  useEffect(() => {
    if (isSubmit) {
      handleSubmit();
    }
  }, [isSubmit]);

  const { mutate, isLoading } = useMutation(postPrescription, {
    onSuccess({ result }) {
      if (result === "success") {
        handleModal(<ConfirmMessage text="처방전 전송이 완료되었습니다." />);
      }
    },
  });

  if (isLoading) {
    return <Loading />;
  }

  const handleChange = (event) => {
    const { id, checked } = event.target;

    if (id === "all") {
      let allCheckedList = doseTimeList.map((time) => {
        return { ...time, isChecked: checked };
      });

      setDoseTimeList(allCheckedList);
    } else {
      let checkedList = doseTimeList.map((time) =>
        time.id === id ? { ...time, isChecked: checked } : time,
      );

      setDoseTimeList(checkedList);
    }
  };

  const handleSubmit = () => {
    const doseTimes = {};

    doseTimeList.forEach(({ isChecked, id }) => {
      if (isChecked) {
        if (id === "beforeBed") {
          return doseTimes["before_bed"] = true;
        }

        doseTimes[id] = true;
      }
    });

    const prescriptionForm = {
      doseTimes,
      medicines: medicineList,
      duration: formData.duration,
      description: formData.description,
      date: new Date().toISOString(),
      patient_id: selectedUser.patient_id,
    };

    if (!selectedUser.name) {
      setError("대기 환자를 선택해주세요.");
      setIsSubmit(false);
      return;
    }

    if (!medicineList.length) {
      setError("처방할 약을 추가해주세요.");
      setIsSubmit(false);
      return;
    }

    if (!Object.keys(doseTimes).length) {
      setError("복용 시간을 체크해주세요.");
      setIsSubmit(false);
      return;
    }

    if (!formData.description) {
      setError("복약지도는 필수 항목입니다.");
      setIsSubmit(false);
      return;
    }

    if (!formData.duration) {
      setError("복용기간을 입력해주세요.");
      setIsSubmit(false);
      return;
    }

    mutate(prescriptionForm);

    setError("");
    setFormData({});
    setIsSubmit(false);
    setSelectedUser({});
    setMedicineList([]);
    setDoseTimeList(doseTimeDataList);
  };

  const doseTimeDataList = [
    { id: "morning", label: "아침", name: "doseTime" },
    { id: "lunch", label: "점심", name: "doseTime" },
    { id: "dinner", label: "저녁", name: "doseTime" },
    { id: "beforeBed", label: "취침전", name: "doseTime" },
  ];

  return (
    <PageContent>
      <PageTitle>Prescription</PageTitle>

      <BoxWrapper>
        <FlexBox>
          <div>
            <UserInfo selectedUser={selectedUser} isSubmit={isSubmit} />
            <SearchForm
              setError={setError}
              medicineList={medicineList}
              setMedicineList={setMedicineList}
            />
          </div>

          <div>
            <CheckboxList>
              <Checkbox
                id="all"
                name="all"
                label="전체"
                checked={doseTimeList.filter(time => time?.isChecked !== true).length < 1}
                onChange={handleChange}
              />
              {doseTimeList.map((time) => {
                const { id, name, label } = time;

                return (
                  <Checkbox
                    key={id}
                    checked={time?.isChecked || false}
                    id={id}
                    name={name}
                    label={label}
                    onChange={handleChange}
                  />
                );
              })}
            </CheckboxList>
          </div>
        </FlexBox>

        <InputForm
          isSubmit={isSubmit}
          setIsSubmit={setIsSubmit}
          setFormData={setFormData}
        />

        {error && <ErrorMessage>{error}</ErrorMessage>}
      </BoxWrapper>

      <BoxWrapper>
        <Queue
          isSubmit={isSubmit}
          selectedUser={selectedUser}
          setSelectedUser={setSelectedUser}
        />
      </BoxWrapper>
    </PageContent>
  );
};

const BoxWrapper = styled.div`
  margin-top: 20px;
  padding: 20px 20px 35px;
  border-radius: 10px;
  background-color: ${({ theme }) => theme.color.white};
  box-shadow: rgb(149 157 165 / 20%) 0px 8px 24px;
`;

const CheckboxList = styled.ul`
  display: inline-flex;
  flex-wrap: wrap;
  justify-content: flex-start;
  max-width: 430px;
  padding: 0 20px;
  vertical-align: top;
  box-sizing: border-box;

  li:first-child {
    border-right: 1px solid #eee;
  }
`;

export default Prescription;
