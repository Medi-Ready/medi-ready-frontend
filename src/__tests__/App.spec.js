import React from "react";
import { screen } from "@testing-library/react";

import App from "../App";
import History from "../pages/History";
import Prescription from "../pages/Prescription";
import { customRender, history } from "../setupTests";
import PrivateRoute from "../components/Shared/PrivateRoute";

describe("<App />", () => {
  it("matches snapshot", () => {
    const { container } = customRender(<App />);

    expect(container).toMatchSnapshot();
  });

  it("should be render login page if user is not loggined in", () => {
    customRender(<App />);
    history.push("/");

    expect(screen.getByText("Google Login")).toBeInTheDocument();
  });

  it("should be render each component matching correct path", () => {
    customRender(<History />);
    history.push("/history");

    expect(screen.getByText("Prescription History")).toBeInTheDocument();
  });

  it("should go to private path when user logged in", () => {
    const user = {
      created_at: "2021-08-31T18:07:09.000Z",
      email: "test@gmail.com",
      exp: 1633002544,
      iat: 1632397744,
      name: "전예림",
      notification_token: null,
      pharmacy_address: "12345 dfghj, Lake Jeremiemouth",
      pharmacy_name: "kenken pharmacy",
      picture: "https://test.com",
      user_id: "test-id",
      user_type: "pharmacist",
    };

    customRender(
      <PrivateRoute
        path="/prescription"
        isAuthenticated={user}
        component={Prescription}
        userInfo={user}
      />,
    );

    history.push("/prescription");

    expect(screen.getByText("처방")).toBeInTheDocument();
    expect(screen.getByText("대기 환자를 선택해주세요.")).toBeInTheDocument();
  });
});
