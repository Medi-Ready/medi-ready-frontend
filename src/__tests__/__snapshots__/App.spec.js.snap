// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> matches snapshot 1`] = `
<div>
  <div
    class="sc-jgHCSr eZVpRi"
  >
    <nav
      class="sc-iBPTik vJucp"
    >
      <h1
        class="sc-fubCzh cjkYFv"
      >
        <a
          href="/"
        >
          MEDI-READY
        </a>
      </h1>
      <nav>
        <ul>
          <li>
            <a
              href="/"
            >
              Dashboard
            </a>
          </li>
          <li>
            <a
              href="/prescription"
            >
              Prescription
            </a>
          </li>
          <li>
            <a
              href="/prescriptions"
            >
              History
            </a>
          </li>
          <li>
            <a
              href="/settings"
            >
              Settings
            </a>
          </li>
          <li>
            <a
              href="/login"
            >
              Login
            </a>
          </li>
        </ul>
      </nav>
    </nav>
    <section
      class="sc-gTgzoy cljoXS"
    >
      <header
        class="sc-hKgJUU iqnQip"
      >
        <div />
      </header>
      <article
        class="sc-kEjbQP eapVfj sc-kstqJO fLJyqo"
      >
        <div>
          <h2
            class="sc-jrAFXE kdvlLA sr-only"
          >
            Login
          </h2>
          <strong>
            Welcome back!
          </strong>
          <p>
            Sign in to your account to continue
          </p>
          <div
            class="sc-hBEYId jTxJXo"
          >
            <button
              class="sc-pGacB gkyJzY sc-fodVek hRwCGu"
              type="button"
            >
              Google Login
            </button>
          </div>
        </div>
      </article>
      <footer
        class="sc-gKseQn jHPEFO"
      >
        2021 Â© Medi-Ready - medi-ready.com
      </footer>
    </section>
  </div>
</div>
`;
